#  **Taller Proceso ETL Aplicado a Bisness Analytics**

**Dataset:** Online Retail

**Lenguaje:** Python

**Fuente de Datos:** UCI Machine Learning Repository

**Presentado por:** Brayan Nicolas Robayo Baquero

**Asignatura:** Business Analytics

***

# **Introducción**

En este trabajo se implementará un proceso ETL utilizando Python y Pandas sobre un conjunto de transacciones comerciales denominado Online Retail. El proceso permitirá trabajar con los datos desde su obtención inicial hasta la generación de un archivo transformado y preparado para análisis.

En este Notebook se documentarán las diferentes etapas, incluyendo la Extracción de los datos, la exploración inicial, la identificación de problemas de calidad, la Transformación y limpieza de la información, y finalmente la carga de los datos procesados.

***

# **Contexto del Proceso ETL**

Este proceso se utiliza para obtener información desde una o varias fuentes, modificarla para mejorar su calidad y estructura, y posteriormente almacenarla en un destino donde pueda ser utilizada para análisis o procesos empresariales.

El proceso se divide en tres etapas principales:

- **Extracción**

Consiste en obtener los datos desde una fuente de origen. Estas fuentes pueden ser archivos, bases de datos, sistemas empresariales, APIs u otros medios de almacenamiento.

- **Transformación**

En esta etapa los datos obtenidos son revisados y procesados para mejorar su calidad y facilitar su utilización. Entre las actividades que pueden realizarse se encuentran la identificación de valores faltantes, eliminación de registros duplicados, modificación de tipos de datos, filtrado de información y creación de nuevas variables.

- **Carga**

La carga consiste en almacenar los datos después de haber sido transformados en un destino adecuado para su utilización. En este proyecto, los datos procesados serán almacenados en un archivo CSV que podrá utilizarse posteriormente para procesos de análisis.

***

# **Referencia Conceptual**

Para la elaboración de este taller se tomo como referencia el Artículo de "Extracción, transformación y carga de datos (ETL) de Microsoft Learn, el cual explica el proceso ETL y su utilización para preparar datos provenientes de distintas fuentes para su posterior análisis"

**Fuente**

Microsoft Learn. Extracción, transformación y carga de datos (ETL).

https://learn.microsoft.com/es-es/azure/architecture/data-guide/relational-data/etl

***

# **Fuente del Conjunto de Datos**

Se selecciono el conjunto de datos Online Retail, disponible en el repositorio UCI Machine Learning Repository, plataforma que comparte conjuntos de datos destinados a actividades de investigación y aprendizaje automático.

La fuente original del conjunto de datos es:

UCI Machine Learning Repository — Online Retail

https://archive.ics.uci.edu/dataset/352/online%2Bretail

***

# **Descripción del Conjunto de Datos**

El archivo utulizado para este taller se encuentra en formato Excel y contiene 541.909 registros y 8 variables.

Este conjunto de datos es apropiado para el desarrollo del proceso ETL porque contiene información real que requiere diferentes procesos de revisión, limpieza y transformación antes de ser utilizada para análisis empresarial.

***

# **Variables del conjunto de datos**

| Variable | Descripción |
|:--------:|:------------|
| InvoiceNo | Número que identifica la factura o transacción. |
| StockCode | Código que identifica el producto. |
| Description | Descripción o nombre del producto. |
| Quantity | Cantidad de unidades registradas en la transacción. |
| InvoiceDate | Fecha y hora en la que se realizó la transacción. |
| UnitPrice | Precio unitario del producto. |
| CustomerID | Identificador del cliente asociado a la transacción. |
| Country | País relacionado con la transacción. |

# **Justificación del Dataset**

Se escogio este conjunto de datos Online Retail porque contiene información relacionada con actividades empresariales de ventas y transacciones comerciales.


De igual forma permite desarrollar las etapas ETL de manera práctica. La variedad de variables permite trabajar con información de productos, clientes, cantidades, precios, fechas, etc. Demostrando como el proceso ETL puede convertir datos originales en información mas estructurada para Business Analytics.

***

# **Etapa 1: Extracción de Datos**

En esta etapa se realizarán las siguientes actividades:

- Importar las bibliotecas necesarias.
- Cargar el archivo Excel.
- Almacenar la información en un DataFrame.
- Comprobar que los datos hayan sido cargados correctamente.
- Realizar una exploración inicial de la información.

***

In [1]:
# Importar librerias para el desarrollo del taller
import pandas as pd # Libreria para trabajar con Datasets
import numpy as np # Libreria para trabajar con datos

In [2]:
# Cargar archivo original
df_retail = pd.read_excel("Online Retail.xlsx") # El archivo se almacenara en un DataFrame

In [3]:
# Ver que se cargo correctamente el DataFrame
df_retail.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [4]:
# Obtener la cantidad de filas y columnas
print(f"Numero de filas: {df_retail.shape[0]} - Numero de columnas: {df_retail.shape[1]}")

Numero de filas: 541909 - Numero de columnas: 8


In [5]:
# Conocer el Dataset - Identificar propiedades de los datos
df_retail.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


# **Análisis de la calidad de los Datos**

**Valores Faltantes**

***


In [6]:
df_retail.isnull().sum() # Buscar valores faltantes por columnas y sumarlos

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0


**Porcentaje de Valores Faltantes**

***

In [7]:
porcentaje_nulos = (df_retail.isnull().sum() / len(df_retail)) * 100 # Obtener porcentaje de los datos faltantes por columnas
np.around(porcentaje_nulos.sort_values(ascending = False), 2) # Ordenar de mayor a menor

,0
CustomerID,24.93
Description,0.27
StockCode,0.00
InvoiceNo,0.00
Quantity,0.00
InvoiceDate,0.00
UnitPrice,0.00
Country,0.00


**Registros Duplicados**

***

In [8]:
duplicados = df_retail.duplicated().sum() # Obtener cantidad de registros duplicados
print(f"Cantidad de registros duplicados: {duplicados}")

Cantidad de registros duplicados: 5268


**Cantidades Negativas**

***

In [9]:
cantidad_negativa = (df_retail['Quantity'] < 0).sum() # Obtener cantidad de registros con cantidades negativas
print(f"Cantidad de registros con cantidades negativas: {cantidad_negativa}")

Cantidad de registros con cantidades negativas: 10624


**Revisión de Precios **

***

In [10]:
precios = (df_retail['UnitPrice'] <= 0).sum() # Obtener cantidad de registros con precios negativos o iguales a 0
print(f"Cantidad de registros con precios negativos: {precios}")

Cantidad de registros con precios negativos: 2517


# **Interpretación de Resultados**

Se encontrarón 5.268 registros duplicados, estos contienen información repetida que pueden generar un conteo incorrecto de las transacciones y afectar calculos.

Se identificarón 10.624 registros con cantidades negativas en la variable Quantity. No representan necesariamente errrores, porque pueden estar relacionados a devoluciones o cancelaciones de productos.

Y finalmente se encontraron 2.517 registros con precios menores o iguales a 0 en la variable UnitPrice. Estos registros deben ser revisados antes de definir si se eliminan o conservan.

Por otro lado hay valores faltantes en variables como lo son CustomerID y description.

Se puede concluir que el conjunto de datos requiere diferentes procesos de limpieza y transformación antes de ser utilizado para análisis empresarial.

***

# **Análisis de Cantidades Negativas**

Antes de eliminar estos registros, se debe analizar el comportamiento de estos, ya que puede representar una devolución  o cancelación, no precisamente un error.

Se revisara la información asociada a estos registros, especialmente el # de factura, la cantidad, precio y descripción del producto.

***

In [11]:
df_retail[df_retail['Quantity'] < 0].head(10) # 10 Primeros registros donde Quantity es negativa

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [12]:
# Comprobar si todas las cantidades negativas tienen factura con C
registros_negativos = df_retail[df_retail['Quantity'] < 0]

# Comprobar si el # de factura empieza por C
facturas_con_c = registros_negativos['InvoiceNo'].astype(str).str.startswith("C")

# Contar cunatos registros cumplen y cuantos no
facturas_con_c.value_counts()

,count
InvoiceNo,
True,9288
False,1336


**Revisión de registros negativos sin factura de cancelación**

Aunque la mayoría de los registros con cantidades negativas corresponden a facturas cuyo número comienza por `C`, también se encontraron 1.336 registros negativos que no presentan esta característica.

Es por esto que, se revisarán estos registros antes de definir el tratamiento que tendrán durante la transformación. Esto permite evitar la eliminación de información que pueda tener algún significado dentro del conjunto de datos.

***

In [13]:
registros_negativos_sin_c = registros_negativos[
    ~registros_negativos["InvoiceNo"].astype(str).str.startswith("C") # El simbolo ~ busca el contrario, en este caso los que no tienen una C
]

registros_negativos_sin_c.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,2010-12-02 14:42:00,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7192,537000,21414,NaN,-22,2010-12-03 15:32:00,0.0,NaN,United Kingdom
7193,537001,21653,NaN,-6,2010-12-03 15:33:00,0.0,NaN,United Kingdom
7195,537003,85126,NaN,-2,2010-12-03 15:33:00,0.0,NaN,United Kingdom
7196,537004,21814,NaN,-30,2010-12-03 15:34:00,0.0,NaN,United Kingdom
7197,537005,21692,NaN,-70,2010-12-03 15:35:00,0.0,NaN,United Kingdom


**Revisión del precio de los registros negativos sin factura C**

Después de identificar los registros con cantidades negativas que no tienen una factura iniciada en `C`, se revisará su precio unitario.

Esta revisión permitirá determinar si estos registros corresponden a operaciones con características diferentes a una venta normal. En particular, se verificará cuántos de estos registros presentan un precio unitario igual a cero.

***

In [14]:
# Contar cuántos registros negativos sin factura C tienen un precio igual a cero
precio_cero = (registros_negativos_sin_c['UnitPrice'] == 0).sum()

print(f"Cantidad de registros con precios iguales a cero: {precio_cero}")

Cantidad de registros con precios iguales a cero: 1336


**Revisión de los registros negativos con precio cero**

Los 1.336 registros con cantidades negativas que no tienen una factura iniciada en `C` presentan un precio unitario igual a cero.

Se revisarán también las variables `Description` y `CustomerID`. Esto permitirá determinar si estos registros contienen información suficiente para considerarlos transacciones comerciales normales o si presentan varias características de datos incompletos.

***

In [15]:
# Contar valores faltantes en Description y CustomerID, dentro de los registros negativos que no tienen factura C
registros_negativos_sin_c[["Description", "CustomerID"]].isnull().sum()

,0
Description,862
CustomerID,1336


**Análisis de precios iguales o inferiores a cero**

Durante el análisis inicial se identificaron 2.517 registros con precios menores o iguales a cero en la variable `UnitPrice`.

Los precios iguales a cero pueden corresponder a situaciones particulares dentro del conjunto de datos y no necesariamente representan un error. Antes de eliminarlos se revisará su relación con otras variables, especialmente `Quantity` e `InvoiceNo`.

El objetivo es determinar si estos registros corresponden a transacciones comerciales normales o si presentan características que justifiquen su exclusión durante la transformación.

***

In [16]:
# Seleccionar los registros cuyo precio unitario es menor o igual a cero
registros_precio_no_valido = df_retail[df_retail["UnitPrice"] <= 0]

# Mostrar los primeros 10 registros para revisar sus características
registros_precio_no_valido.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom


**Diferenciación entre precios iguales a cero y precios negativos**

En el análisis anterior se identificaron 2.517 registros cuyo `UnitPrice` es menor o igual a cero. Sin embargo, esta condición incluye dos situaciones diferentes: precios exactamente iguales a cero y precios negativos.

Se separarán ambos casos para conocer su cantidad y determinar posteriormente el tratamiento adecuado dentro del proceso de transformación.

***

In [17]:
# Contar registros con precio exactamente igual a cero
precios_cero = (df_retail["UnitPrice"] == 0).sum()

# Contar registros con precio negativo
precios_negativos = (df_retail["UnitPrice"] < 0).sum()

# Mostrar los resultados
print("Registros con precio igual a cero:", precios_cero)
print("Registros con precio negativo:", precios_negativos)

Registros con precio igual a cero: 2515
Registros con precio negativo: 2


**Revisión de registros con precio negativo**

De los 2.517 registros con precios menores o iguales a cero, se identificaron únicamente 2 registros con un precio estrictamente negativo.

Debido a que un precio negativo no representa normalmente un precio de venta, estos registros serán revisados individualmente antes de definir su tratamiento durante la transformación.

***

In [18]:
# Filtrar únicamente los registros cuyo precio unitario es negativo
registros_precio_negativo = df_retail[df_retail["UnitPrice"] < 0]

# Mostrar todos los registros encontrados
registros_precio_negativo

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


**Interpretación de los precios negativos**

Los 2 registros con precios negativos presentan características diferentes a las de una venta normal. La descripción corresponde a `Adjust bad debt`, la cantidad registrada es 1 y el precio unitario es de -11062.06.

Estos registros corresponden a ajustes relacionados con deuda y no representan una venta de productos. Por esta razón, durante la etapa de transformación serán excluidos del conjunto de datos destinado al análisis de transacciones comerciales.

Esta decisión se toma teniendo en cuenta el objetivo del proceso ETL, que es preparar los datos de ventas para su posterior análisis empresarial.

***

**Análisis de registros con precio igual a cero**

Después de revisar los precios negativos, se analizarán los 2.515 registros cuyo `UnitPrice` es igual a cero.

Es necesario revisar estos registros según el signo de `Quantity`, ya que una cantidad positiva puede representar una operación diferente a una devolución o cancelación.

Se separarán los registros con cantidades positivas, negativas y cero para comprender mejor su comportamiento antes de aplicar una transformación.

***

In [19]:
# Seleccionar los registros cuyo precio unitario es exactamente cero
registros_precio_cero = df_retail[df_retail["UnitPrice"] == 0]

# Contar registros con cantidad positiva
cantidad_positiva = (registros_precio_cero["Quantity"] > 0).sum()

# Contar registros con cantidad negativa
cantidad_negativa = (registros_precio_cero["Quantity"] < 0).sum()

# Contar registros con cantidad igual a cero
cantidad_cero = (registros_precio_cero["Quantity"] == 0).sum()

# Mostrar los resultados
print("Registros con Quantity positiva:", cantidad_positiva)
print("Registros con Quantity negativa:", cantidad_negativa)
print("Registros con Quantity igual a cero:", cantidad_cero)

Registros con Quantity positiva: 1179
Registros con Quantity negativa: 1336
Registros con Quantity igual a cero: 0


**Revisión de cantidades positivas con precio cero**

De los 2.515 registros con precio unitario igual a cero, se encontraron 1.179 registros que tienen una cantidad positiva.

Estos registros son diferentes de los 1.336 registros negativos analizados anteriormente. Por esta razón, se revisarán sus características para determinar si corresponden a operaciones comerciales válidas o a registros que no deberían formar parte del conjunto de datos utilizado para el análisis de ventas.

Se revisarán principalmente el número de factura, la descripción del producto, la cantidad y el cliente asociado.

***

In [20]:
# Seleccionar los registros que tienen precio cero y cantidad positiva
registros_precio_cero_positivos = registros_precio_cero[registros_precio_cero["Quantity"] > 0]

# Mostrar los primeros 10 registros para revisar sus características
registros_precio_cero_positivos.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
4348,536765,84952C,NaN,19,2010-12-02 14:43:00,0.0,NaN,United Kingdom


**Revisión de valores faltantes en registros con precio cero**

En la muestra de los registros con cantidad positiva y precio unitario igual a cero se observan valores faltantes en las variables `Description` y `CustomerID`.

Se calculará la cantidad de valores faltantes en estas dos variables dentro de los 1.179 registros identificados.

***

In [21]:
# Revisar cuántos valores faltantes existen en Description y CustomerID
# dentro de los registros con cantidad positiva y precio igual a cero

registros_precio_cero_positivos[["Description", "CustomerID"]].isnull().sum()

,0
Description,592
CustomerID,1139


**Revisión del tipo de factura**

Los registros con cantidad positiva y precio igual a cero presentan diferentes niveles de información faltante. Antes de decidir su tratamiento se revisará el número de factura.

Se comprobará cuántos de estos registros tienen una factura cuyo número comienza por `C` y cuántos corresponden a facturas que no comienzan por `C`.

Esta revisión permitirá identificar si existen características diferentes dentro de este grupo de registros.

***

In [22]:
# Comprobar si las facturas de los registros con precio cero y cantidad positiva comienzan con la letra C

facturas_c_precio_cero = (
    registros_precio_cero_positivos["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

# Contar las facturas que comienzan con C y las que no
facturas_c_precio_cero.value_counts()

,count
InvoiceNo,
False,1179


**Revisión de registros con precio cero y descripción disponible**

Dentro de los 1.179 registros con cantidad positiva y precio igual a cero, se identificaron registros con y sin descripción.

Se revisarán algunos registros que sí tienen una descripción del producto. Esto permitirá determinar si los precios iguales a cero corresponden a productos u operaciones particulares que pueden tener sentido dentro del conjunto de datos.

***

In [23]:
# Seleccionar únicamente los registros con precio cero,
# cantidad positiva y una descripción disponible
registros_cero_con_descripcion = registros_precio_cero_positivos[
    registros_precio_cero_positivos["Description"].notnull()
]

# Mostrar los primeros 10 registros
registros_cero_con_descripcion.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,NaN,United Kingdom
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647.0,Germany
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14338,537534,48184,DOORMAT ENGLISH ROSE,3,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14339,537534,48111,DOORMAT 3 SMILEY CATS,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14340,537534,22697,GREEN REGENCY TEACUP AND SAUCER,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom
14341,537534,22682,FRENCH BLUE METAL DOOR SIGN 7,1,2010-12-07 11:48:00,0.0,NaN,United Kingdom


**Identificación de las descripciones más frecuentes**

Los registros con precio unitario igual a cero y cantidad positiva contienen diferentes productos y operaciones.

Se identificarán las descripciones que aparecen con mayor frecuencia. Esto permitirá detectar si los precios cero están asociados principalmente a determinados productos u operaciones específicas.

***

In [24]:
# Contar cuántas veces aparece cada descripción
# dentro de los registros con precio cero y cantidad positiva
frecuencia_descripciones = (
    registros_cero_con_descripcion["Description"]
    .value_counts()
)

# Mostrar las 15 descripciones más frecuentes
frecuencia_descripciones.head(15)

,count
Description,
check,39
found,25
adjustment,14
FRENCH BLUE METAL DOOR SIGN 1,9
FRENCH BLUE METAL DOOR SIGN 8,8
amazon,8
Found,8
FRENCH BLUE METAL DOOR SIGN 3,7
RECIPE BOX PANTRY YELLOW DESIGN,7


# **Transformación y limpieza de los datos**

En esta etapa se aplicarán diferentes procedimientos para mejorar la calidad y estructura de los datos. Las transformaciones se realizarán de manera progresiva, teniendo en cuenta los problemas identificados durante el análisis anterior.

***

**Eliminación de registros duplicados**

Durante el análisis inicial se identificaron 5.268 registros duplicados.

Los registros duplicados contienen exactamente los mismos valores en todas sus columnas. Mantenerlos podría provocar que determinadas operaciones fueran contabilizadas más de una vez y afectar cálculos posteriores.

Por esta razón, se eliminarán los registros completamente duplicados, conservando únicamente una copia de cada registro.

In [25]:
# Guardar la cantidad de registros antes de eliminar duplicados
registros_antes = len(df_retail)

# Eliminar los registros que están completamente duplicados
df_retail = df_retail.drop_duplicates()

# Guardar la cantidad de registros después de eliminar duplicados
registros_despues = len(df_retail)

# Calcular cuántos registros fueron eliminados
duplicados_eliminados = registros_antes - registros_despues

# Mostrar los resultados
print("Registros antes de eliminar duplicados:", registros_antes)
print("Registros después de eliminar duplicados:", registros_despues)
print("Registros duplicados eliminados:", duplicados_eliminados)

Registros antes de eliminar duplicados: 541909
Registros después de eliminar duplicados: 536641
Registros duplicados eliminados: 5268


**Tratamiento de cancelaciones y devoluciones**

Durante el análisis inicial se identificaron 10.624 registros con cantidades negativas.

La revisión realizada mostró que 9.288 de estos registros tienen un número de factura que comienza con la letra `C`, característica utilizada en este conjunto de datos para identificar cancelaciones.

Para este proceso ETL se busca construir un conjunto de datos orientado al análisis de las ventas realizadas. Por esta razón, las transacciones identificadas como cancelaciones serán excluidas del conjunto de datos final.

También se excluirán los registros con cantidades negativas que no presentan una factura iniciada en `C`, ya que durante el análisis se encontró que estos registros presentan características diferentes a una venta normal, como precio unitario igual a cero y ausencia de información del cliente.

Antes de aplicar la eliminación se calculará la cantidad de registros que serán afectados para poder verificar posteriormente el resultado.

***

In [26]:
# Identificar las facturas que comienzan con la letra C
# Estas corresponden a cancelaciones dentro del conjunto de datos
es_cancelacion = (
    df_retail["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

# Identificar los registros que tienen cantidad negativa
es_cantidad_negativa = df_retail["Quantity"] < 0

# Contar cuántos registros cumplen cada condición
print("Registros con factura C:", es_cancelacion.sum())
print("Registros con cantidad negativa:", es_cantidad_negativa.sum())

# Contar cuántos registros cumplen al menos una de las dos condiciones
registros_a_excluir = es_cancelacion | es_cantidad_negativa

print("Registros que cumplen alguna de las condiciones:",
      registros_a_excluir.sum())

Registros con factura C: 9251
Registros con cantidad negativa: 10587
Registros que cumplen alguna de las condiciones: 10587


**Eliminación de cancelaciones y registros con cantidades negativas**

Después de analizar los registros con cantidades negativas, se determinó que todos los registros correspondientes a facturas iniciadas en `C` también presentan cantidades negativas.

Por lo tanto, para preparar un conjunto de datos orientado al análisis de ventas, se utilizará `Quantity < 0` como criterio para excluir las cancelaciones, devoluciones y demás registros con cantidades negativas.

Esta transformación permitirá trabajar posteriormente con las transacciones que presentan cantidades positivas.

***

In [27]:
# Guardar la cantidad de registros antes de eliminar las cantidades negativas
registros_antes_negativos = len(df_retail)

# Conservar únicamente los registros con cantidades positivas
df_retail = df_retail[df_retail["Quantity"] > 0].copy()

# Guardar la cantidad de registros después de la transformación
registros_despues_negativos = len(df_retail)

# Calcular cuántos registros fueron eliminados
registros_negativos_eliminados = (
    registros_antes_negativos - registros_despues_negativos
)

# Mostrar los resultados
print(
    "Registros antes de eliminar cantidades negativas:",
    registros_antes_negativos
)

print(
    "Registros después de la transformación:",
    registros_despues_negativos
)

print(
    "Registros eliminados:",
    registros_negativos_eliminados
)

Registros antes de eliminar cantidades negativas: 536641
Registros después de la transformación: 526054
Registros eliminados: 10587


**Tratamiento de precios negativos**

Durante el análisis inicial se identificaron únicamente 2 registros con precios negativos.

Al revisar estos registros se encontró que ambos tienen la descripción `Adjust bad debt` y corresponden a ajustes relacionados con deuda, por lo que no representan ventas de productos.

Por esta razón, estos 2 registros serán excluidos del conjunto de datos utilizado para el análisis de ventas.

Los registros con `UnitPrice = 0` no serán eliminados en esta etapa, debido a que el análisis mostró que algunos corresponden a productos con información de transacción y podrían representar operaciones especiales.

***

In [28]:
# Guardar la cantidad de registros antes de eliminar precios negativos
registros_antes_precio = len(df_retail)

# Conservar únicamente los registros cuyo precio sea mayor o igual a cero
df_retail = df_retail[df_retail["UnitPrice"] >= 0].copy()

# Guardar la cantidad de registros después de la transformación
registros_despues_precio = len(df_retail)

# Calcular cuántos registros fueron eliminados
precios_negativos_eliminados = (
    registros_antes_precio - registros_despues_precio
)

# Mostrar los resultados
print(
    "Registros antes de eliminar precios negativos:",
    registros_antes_precio
)

print(
    "Registros después de la transformación:",
    registros_despues_precio
)

print(
    "Precios negativos eliminados:",
    precios_negativos_eliminados
)

Registros antes de eliminar precios negativos: 526054
Registros después de la transformación: 526052
Precios negativos eliminados: 2


**Revisión de valores faltantes**

Después de eliminar los registros duplicados, las cantidades negativas y los precios negativos, es necesario volver a revisar los valores faltantes.

No todos los valores faltantes deben eliminarse automáticamente. Por ejemplo, la ausencia de `CustomerID` no significa necesariamente que la transacción sea inválida, ya que la venta puede contener información útil sobre el producto, la cantidad, el precio y el país.

***

In [29]:
# Contar los valores faltantes de cada columna
valores_faltantes_actuales = df_retail.isnull().sum()

# Mostrar únicamente las columnas que tienen valores faltantes
valores_faltantes_actuales[valores_faltantes_actuales > 0]

,0
Description,592
CustomerID,133320


**Tratamiento de valores faltantes**

Después de revisar nuevamente los datos se encontraron valores faltantes en las variables `Description` y `CustomerID`.

Los valores faltantes de `Description` serán reemplazados por el texto "Sin descripción", ya que de esta manera se conserva el registro de la transacción sin inventar información sobre el producto.

En el caso de `CustomerID`, los valores faltantes se conservarán como datos no identificados. No se eliminarán estos registros porque la ausencia del identificador del cliente no significa que la transacción sea inválida y eliminar estas filas ocasionaría una pérdida importante de información para el análisis de ventas.

De esta manera, se busca conservar la mayor cantidad posible de información original y evitar generar datos que no se encuentran en la fuente.

***

In [30]:
# Reemplazar las descripciones faltantes por un texto que indique que no están disponibles
df_retail["Description"] = df_retail["Description"].fillna("Sin descripción")

# Contar nuevamente los valores faltantes
valores_faltantes = df_retail.isnull().sum()

# Mostrar únicamente las columnas que todavía tienen valores faltantes
valores_faltantes[valores_faltantes > 0]

,0
CustomerID,133320


**Ajuste de tipos de datos**

Después de la limpieza de los valores faltantes, se revisarán los tipos de datos de las variables.

Se verificará que las variables relacionadas con cantidades, precios y fechas mantengan los tipos adecuados para realizar posteriormente cálculos y análisis.

***

In [31]:
# Mostrar los tipos de datos actuales de cada columna
df_retail.dtypes

,0
InvoiceNo,object
StockCode,object
Description,object
Quantity,int64
InvoiceDate,datetime64[ns]
UnitPrice,float64
CustomerID,float64
Country,object


In [32]:
# Convertir CustomerID a entero, manteniendo los valores faltantes
df_retail["CustomerID"] = df_retail["CustomerID"].astype("Int64")

# Mostrar nuevamente los tipos de datos
df_retail.dtypes

,0
InvoiceNo,object
StockCode,object
Description,object
Quantity,int64
InvoiceDate,datetime64[ns]
UnitPrice,float64
CustomerID,Int64
Country,object


**Creación de la variable de valor total**

Se creará una nueva variable llamada `TotalVenta`.

Esta variable se calculará multiplicando la cantidad de productos vendidos (`Quantity`) por el precio unitario (`UnitPrice`).

La fórmula utilizada será:

TotalVenta = Quantity × UnitPrice

Esta nueva variable permitirá conocer el valor económico asociado a cada registro y posteriormente facilitará análisis como ventas por producto, país, cliente o periodo.

***

In [33]:
# Calcular el valor total de cada registro de venta
df_retail["TotalVenta"] = (
    df_retail["Quantity"] * df_retail["UnitPrice"]
)

# Mostrar algunos registros para comprobar el cálculo
df_retail[["Quantity", "UnitPrice", "TotalVenta"]].head()

,Quantity,UnitPrice,TotalVenta
0,6,2.55,15.30
1,6,3.39,20.34
2,8,2.75,22.00
3,6,3.39,20.34
4,6,3.39,20.34


**Creación de variables relacionadas con la fecha**

La variable `InvoiceDate` contiene tanto la fecha como la hora de cada transacción.

Para facilitar futuros análisis empresariales, se crearán nuevas variables a partir de esta información. Estas variables permitirán analizar las ventas por año, mes y día sin tener que procesar nuevamente la fecha original.

Se crearán las variables `Año`, `Mes` y `Día`.

***

In [34]:
# Extraer el año de cada transacción
df_retail["Año"] = df_retail["InvoiceDate"].dt.year

# Extraer el mes de cada transacción
df_retail["Mes"] = df_retail["InvoiceDate"].dt.month

# Extraer el día de cada transacción
df_retail["Día"] = df_retail["InvoiceDate"].dt.day

# Mostrar las variables creadas junto con la fecha original
df_retail[["InvoiceDate", "Año", "Mes", "Día"]].head()

,InvoiceDate,Año,Mes,Día
0,2010-12-01 08:26:00,2010,12,1
1,2010-12-01 08:26:00,2010,12,1
2,2010-12-01 08:26:00,2010,12,1
3,2010-12-01 08:26:00,2010,12,1
4,2010-12-01 08:26:00,2010,12,1


**Validación de los datos transformados**

Se realizará una validación final para comprobar que las transformaciones aplicadas se realizaron correctamente.

Se verificará la cantidad de registros, los valores faltantes, los duplicados, las cantidades negativas y los precios negativos.

Esta validación permite comprobar que el conjunto de datos cumple con las condiciones definidas durante el proceso de limpieza.

***

In [35]:
# Mostrar la cantidad actual de registros y columnas
print("Cantidad de registros:", len(df_retail))
print("Cantidad de columnas:", len(df_retail.columns))

# Comprobar registros duplicados
print("Registros duplicados:", df_retail.duplicated().sum())

# Comprobar cantidades negativas
print("Cantidades negativas:", (df_retail["Quantity"] < 0).sum())

# Comprobar precios negativos
print("Precios negativos:", (df_retail["UnitPrice"] < 0).sum())

# Comprobar valores faltantes
print("\nValores faltantes:")
print(df_retail.isnull().sum())

Cantidad de registros: 526052
Cantidad de columnas: 12
Registros duplicados: 0
Cantidades negativas: 0
Precios negativos: 0

Valores faltantes:
InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133320
Country             0
TotalVenta          0
Año                 0
Mes                 0
Día                 0
dtype: int64


# **Carga de los datos (Load)**

Una vez finalizadas las etapas de extracción y transformación, se realizará la carga del conjunto de datos procesado.

El resultado será almacenado en un archivo CSV denominado `online_retail_limpio.csv`.

Este archivo contendrá los datos después de aplicar los procesos de limpieza y transformación realizados durante el proyecto, por lo que podrá ser utilizado posteriormente para análisis de Business Analytics.

***

In [36]:
# Nombre del archivo que se generará
archivo_salida = "online_retail_limpio.csv"

# Guardar el DataFrame transformado en formato CSV
df_retail.to_csv(
    archivo_salida,
    index=False,
    encoding="utf-8-sig"
)

# Confirmar que el archivo fue generado
print("Archivo generado correctamente:", archivo_salida)

Archivo generado correctamente: online_retail_limpio.csv


**Verificación del archivo generado**

Después de generar el archivo CSV, se realizará una lectura del archivo para comprobar que los datos fueron almacenados correctamente.

Esta verificación permite confirmar que el archivo contiene la información transformada y que conserva la estructura esperada.

***

In [37]:
# Leer nuevamente el archivo CSV generado
df_verificacion = pd.read_csv(
    "online_retail_limpio.csv"
)

# Mostrar las primeras filas del archivo
df_verificacion.head()

/tmp/ipykernel_2068/4054327391.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df_verificacion = pd.read_csv(


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,TotalVenta,Año,Mes,Día
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,2010,12,1
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,1
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,2010,12,1
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,1
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,1


In [38]:
# Comparar la cantidad de registros antes y después de guardar el CSV
print("Registros del DataFrame transformado:", len(df_retail))
print("Registros del archivo CSV:", len(df_verificacion))

# Comprobar si ambas cantidades son iguales
print(
    "¿La cantidad de registros coincide?:",
    len(df_retail) == len(df_verificacion)
)

Registros del DataFrame transformado: 526052
Registros del archivo CSV: 526052
¿La cantidad de registros coincide?: True


# **Resultado final del proceso ETL**

Después de realizar las etapas de extracción, transformación y carga, se obtuvo un conjunto de datos preparado para posteriores procesos de análisis empresarial.

El conjunto de datos pasó de 541.909 registros originalesa 526.052 registros después de eliminar duplicados, registros con cantidades negativas y registros con precios negativos.

Durante la transformación también se trataron los valores faltantes de la variable `Description`, se conservaron los registros sin `CustomerID`, se creó la variable `TotalVenta` y se generaron las variables `Año`, `Mes` y `Día` a partir de la fecha de cada transacción.

Finalmente, los datos transformados fueron almacenados en el archivo **`online_retail_limpio.csv`** y se comprobó que el archivo contiene los mismos 526.052 registros del DataFrame transformado.

De esta manera, se completó el proceso ETL:

Extracción → Transformación → Validación → Carga

***